src/main/scala/calc/Typer.scala

```diff
@@ -0,0 +1,25 @@
+package calc
+
+object Typer {
+
+  def inferType(tree: Tree): TreeT = {
+    tree match {
+      case t:Literal => implicit val pos = t.pos
+        LiteralT(t.value)
+
+      case t:BinaryOp => binaryOp(t)
+    }
+  }
+
+  def binaryOp(t: BinaryOp) = { implicit val pos = t.pos
+    val lhs = inferType(t.lhs)
+    val rhs = inferType(t.rhs)
+    if (lhs.getType() == rhs.getType()) {
+      new BinaryOpT(t.op, lhs, rhs) {
+        def getType() = lhs.getType()
```

Comment:
Using a `def` here is dangerous from a complexity point of view. If your lhs is itself a `BinaryOpT`, and recursively so, calling a single `getType` will take liner time, and then `binaryOp()` itself will blow up exponentially.